lib/cardinality-check.js

```diff
@@ -0,0 +1,79 @@
+'use strict';
+
+var setInterval = require('timers').setInterval;
+var clearInterval = require('timers').clearInterval;
+var assert = require('assert');
+
+var DEFAULT_INTERVAL_TIME = 10 * 1000;
+var DEFAULT_STAT_NAME = 'uber-statsd-client.total-cardinality';
+var DEFAULT_MAXIMUM_LENGTH = Infinity;
+
+module.exports = CardinalityCheck;
+
+function CardinalityCheck(options) {
+    if (!(this instanceof CardinalityCheck)) {
+        return new CardinalityCheck(options);
+    }
+
+    var self = this;
+
+    assert(options && options.statsdClient,
+        'CardinalityCheck needs options.statsdClient');
+
+    self.statsdClient = options.statsdClient;
+    self.intervalTime = options.intervalTime ||
+        DEFAULT_INTERVAL_TIME;
+    self.statName = options.statName ||
+        DEFAULT_STAT_NAME;
+    self.maximumLength = 'maximumLength' in options ?
+        options.maximumLength : DEFAULT_MAXIMUM_LENGTH;
+
+    self.interval = null;
+    self.knownKeys = Object.create(null);
+    self.knownKeysLength = 0;
+    self.full = false;
+}
+
+CardinalityCheck.prototype.bootstrap = function bootstrap() {
+    var self = this;
+
+    self.interval = setInterval(onReport, self.intervalTime);
+    self.interval.unref();
+
+    function onReport() {
+        self._reportCardinality();
+    }
+};
+
+CardinalityCheck.prototype.addKey = function addKey(key) {
+    var self = this;
+
+    if (self.full) {
+        return;
+    }
```

Comment:
having `knownKeys[key]` be the number of times the stat is emitted is purely for debugging.

When you leak a bunch of stats; far more then you expected, which would be more useful

 -  Getting the number of times the stats have been emitted once we passed the `maximumNumberOfKeys` threshold
 - Getting the number of times the stats have been emitted up until right now.

I guess the latter is more useful because you could in theory do a diff between two moments in time to detect what is growing or *not* growing.